File: widgetssdk/src/main/java/com/glia/widgets/core/engagement/domain/MapOperatorUseCase.java
package com.glia.widgets.core.engagement.domain;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.glia.androidsdk.Operator;
import com.glia.androidsdk.chat.ChatMessage;
import com.glia.androidsdk.chat.OperatorMessage;
import com.glia.widgets.core.engagement.domain.model.ChatMessageInternal;

import io.reactivex.Single;

public class MapOperatorUseCase {
    private final GetOperatorUseCase getOperatorUseCase;

    public MapOperatorUseCase(GetOperatorUseCase getOperatorUseCase) {
        this.getOperatorUseCase = getOperatorUseCase;
    }

    public Single<ChatMessageInternal> execute(ChatMessage chatMessage) {
        if (chatMessage.isOperator()) {
            return Single.just(chatMessage).cast(OperatorMessage.class).flatMap(this::mapOperator);
        } else {
            return Single.just(chatMessage).map(ChatMessageInternal::new);
        }
    }

    @NonNull
    private Single<ChatMessageInternal> mapOperator(OperatorMessage operatorMessage) {
        return getOperatorUseCase.execute(operatorMessage.getOperatorId())
                .map(operator -> map(operatorMessage, operator.orElse(null)));
    }

    @NonNull
    private ChatMessageInternal map(@NonNull OperatorMessage operatorMessage, @Nullable Operator operator) {
        return new ChatMessageInternal(operatorMessage, operator);
    }

}

File: widgetssdk/src/main/java/com/glia/widgets/core/engagement/domain/MapOperatorUseCase.kt
package com.glia.widgets.core.engagement.domain

import com.glia.androidsdk.chat.Chat
import com.glia.androidsdk.chat.ChatMessage
import com.glia.androidsdk.chat.OperatorMessage
import com.glia.widgets.core.engagement.domain.model.ChatMessageInternal
import io.reactivex.Single
import kotlin.jvm.optionals.getOrNull

internal class MapOperatorUseCase(private val getOperatorUseCase: GetOperatorUseCase) {
    operator fun invoke(chatMessage: ChatMessage): Single<ChatMessageInternal> =
        when (chatMessage.senderType) {
            Chat.Participant.OPERATOR -> processOperatorMessage(chatMessage as OperatorMessage)
            else -> processVisitorMessage(chatMessage)
        }

    private fun processOperatorMessage(chatMessage: OperatorMessage): Single<ChatMessageInternal> =
        getOperatorUseCase.execute(chatMessage.operatorId!!).map { ChatMessageInternal(chatMessage, it.getOrNull()) }

    private fun processVisitorMessage(chatMessage: ChatMessage): Single<ChatMessageInternal> = Single.just(ChatMessageInternal(chatMessage))
}
